src/prettyprint.lua

```diff
@@ -0,0 +1,229 @@
+local function printSummary(results)
+	local passedTests, failedTests, totalTests = 0, 0, 0
+	local passedSuites, failedSuites, totalSuites = 0, 0, #results
+
+	for _, testSuite in ipairs(results) do
+		if testSuite.pass then
+			passedSuites = passedSuites + 1
+		else
+			failedSuites = failedSuites + 1
+		end
+	end
+
+	for _, testSuite in ipairs(results) do
+		traverseNodes(testSuite, function(node, _)
+			if node.type == NodeType.Test then
+				totalTests = totalTests + 1
+				if node.pass then
+					passedTests = passedTests + 1
+				else
+					failedTests = failedTests + 1
+				end
+			end
+
+			return true
+		end)
+	end
+
+	--- Creates a colored summary element which omits certain information if it's not necessary to the reader.
+	---@param header string Header to display
+	---@param failed number
+	---@param passed number
+	---@param total number
+	---@return string rendered Rendered display
+	local function createSummaryInfo(header, failed, passed, total)
+		local stringTable = {}
+		if failed > 0 then
+			tablex.push(
+				stringTable,
+				COLOURS.TESTS_FAILED(("%d failed"):format(failed))
+			)
+		end
+
+		if passed > 0 then
+			tablex.push(
+				stringTable,
+				COLOURS.TESTS_PASSED(("%d passed"):format(passed))
+			)
+		end
+
+		tablex.push(stringTable, ("%d total"):format(total))
+		return ("%s %s"):format(
+			COLOURS.BOLD(header),
+			table.concat(stringTable, ", ")
+		)
+	end
+
+	print(
+		createSummaryInfo(
+			"Test Suites:",
+			failedSuites,
+			passedSuites,
+			totalSuites
+		)
+	)
+
+	print(createSummaryInfo("Tests:", failedTests, passedTests, totalTests))
+end
+
+--- Prints detailed error reports about the tests that failed.
+---@param results lest.TestSuiteResults
+local function printTestErrors(results)
+	---@type table<number, {displayName: string, result: lest.TestResult}>
+	local failedTests = {}
+
+	---@type table<number, string>
+	local parentNames = {}
+	for _, testSuite in ipairs(results) do
+		if not testSuite.passed then
+			traverseNodes(testSuite, function(node)
+				if node.type == NodeType.Describe then
+					if node.pass then
+						-- Skip traversing this branch as it doesn't have anything that failed.
+						return false
+					end
+
+					tablex.push(parentNames, node.name)
+				elseif node.type == NodeType.Test and not node.pass then
+					tablex.push(failedTests, {
+						displayName = table.concat(parentNames, " › ")
+							.. " › "
+							.. node.name,
+						result = node,
+					})
+				end
+
+				return true
+			end, function()
+				tablex.pop(parentNames)
+			end)
+		end
+	end
+
+	for _, failedTest in ipairs(failedTests) do
+		print(
+			COLOURS.ERROR_HEADER(
+				string.format(" %s %s\n", ERROR_SYMBOL, failedTest.displayName)
+			)
+		)
+
+		if
+			type(failedTest.result.error) == "table"
+			and failedTest.result.error.signature
+		then
+			print("   " .. failedTest.result.error.signature .. "\n")
+			print(
+				"   " .. COLOURS.FAIL(failedTest.result.error.message) .. "\n"
+			)
+		else
```

Comment:
This printing behaviour should actually be implemented as part of the `TestResultError` in the `__tostring` method. By doing this we can simplify this function to only the else path.

This is actually an instance of feature envy.
> As a rule of thumb, whenever you find yourself making extensive use of fields from another class to perform any sort of logic or computation, consider moving that logic to a method on the class itself.